test/src/test/java/hudson/PluginManagerTest.java

```diff
@@ -533,4 +535,20 @@ public void findResourceForPluginFirstClassLoader() throws Exception {
 
         assertEquals(fromPlugin, fromToolkit);
     }
+
+    // Sources for jenkins-50336.hpi are available at https://github.com/Vlatombe/jenkins-50336
+    //
+    // package io.jenkins.plugins;
+    // import org.jenkinsci.plugins.variant.OptionalExtension;
+    // import jenkins.model.GlobalConfiguration;
+    // @OptionalExtension public class MyGlobalConfiguration extends GlobalConfiguration {}
+    //
+    @Issue("JENKINS-50336")
+    @Test
+    public void optionalExtensionCanBeFoundAfterDynamicLoadOfVariant() throws Exception {
+        dynamicLoad("variant.hpi");
+        assertNotNull(r.jenkins.getPluginManager().getPlugin("variant"));
+        dynamicLoad("jenkins-50336.hpi");
```

Comment:
Yes, but I do not have a demo or anything at this point. I'm not sure what the best way would be to handle custom code in the mock plugins, since writing java code in strings seems less than ideal, but I guess for a case like this it would be totally fine.